File: backend/gemini_key.py
"""
gemini_key.py
=============
Reads a student's own Gemini API key off the request.

The model: browser-held, never stored
-------------------------------------
The key lives in the student's browser (`frontend/src/lib/gemini-key.ts`,
localStorage) and rides along on the requests that need it, in the
`X-Gemini-Api-Key` header. This service uses it for that one call and forgets
it.

Nothing persists it. There is no column for it, no encryption key to manage,
and no table that becomes interesting to an attacker. That is the entire point
of the design: the key has to reach the server either way - Gemini is called
server-side, so a browser-only key is not possible - but at-rest storage is
avoidable, so it is avoided.

The trade the student makes is per-device: the key is in one browser's
localStorage, so they re-paste it on their phone. For a free key that takes
thirty seconds to mint, that is the right side of the trade.

Two rules this file exists to enforce
-------------------------------------
* The key NEVER gets logged. Not in an access log, not in an exception, not in
  a debug line. `ai_service._redact` covers the error paths; this module simply
  never writes the value anywhere.
* The key NEVER gets stored, echoed back, or included in a response. The API
  reports *whether* a key was used, never what it was.
"""

from __future__ import annotations

from typing import Optional

from fastapi import Header, HTTPException, status

# Custom header rather than a query parameter or the body, for three reasons:
# it survives multipart uploads (the tailoring endpoint is multipart), it never
# lands in a URL and therefore never in a server access log or browser history,
# and it is uniform across every endpoint that needs it.
HEADER_NAME = "X-Gemini-Api-Key"

# Google AI Studio keys are currently 39 characters and start with "AIza". The
# bounds below are deliberately wider than that: a hard format check would
# reject a perfectly good key the day Google changes the format, and the real
# validation is Gemini's answer. This only catches the obvious mistakes -
# pasting a truncated key, or a whole URL - so the student gets a useful
# message in milliseconds instead of a generic AI failure after an upload.
MIN_KEY_LENGTH = 20
MAX_KEY_LENGTH = 200


def get_student_api_key(
    x_gemini_api_key: Optional[str] = Header(
        default=None,
        alias=HEADER_NAME,
        description=(
            "The student's own Google Gemini API key. When present, it is used "
            "for this request instead of the app's key and the request does not "
            "count against the free weekly allowance. Never stored."
        ),
    ),
) -> Optional[str]:
    """The student's own key for this request, or None to use the free allowance.

    FastAPI dependency. Returns None for both "header absent" and "header
    present but empty", because a frontend that clears a saved key may well
    send the empty string, and treating that as a malformed key rather than as
    "no key" would be a confusing error for a student who just removed theirs.
    """
    if x_gemini_api_key is None:
        return None

    key = x_gemini_api_key.strip()
    if not key:
        return None

    if len(key) < MIN_KEY_LENGTH or len(key) > MAX_KEY_LENGTH:
        # The message must not echo the value back - see the module docstring.
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=(
                "That does not look like a Gemini API key. Copy the whole key "
                "from Google AI Studio - it starts with 'AIza' and has no "
                "spaces."
            ),
        )

    # Whitespace inside a key means a line break survived the copy-paste, which
    # would otherwise fail as an authentication error and read as "my key is
    # wrong" rather than "my paste is wrong".
    if any(character.isspace() for character in key):
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=(
                "That key contains a space or line break, so it was probably "
                "copied with extra characters. Paste it again."
            ),
        )

    return key

File: frontend/src/lib/gemini-key.ts
/**
 * gemini-key.ts
 * =============
 * The student's own Gemini API key, held in this browser only.
 *
 * The model
 * ---------
 * ResumeMaxxer funds a few tailoring runs a week per student. Past that, the
 * student supplies their own Gemini key - free from Google AI Studio - and
 * their runs come off their own quota instead.
 *
 * The key is stored in `localStorage` and sent on the requests that need it,
 * in the `X-Gemini-Api-Key` header. The server uses it for that one call and
 * forgets it: no column, no encryption key to manage, no table worth stealing.
 *
 * Why not store it server-side, given it reaches the server anyway? Because
 * "reaches the server" and "is kept by the server" are different exposures.
 * Gemini is called server-side, so the key must travel - that is unavoidable.
 * Keeping a copy at rest is avoidable, so it is avoided.
 *
 * What the student trades for that
 * --------------------------------
 * `localStorage` is per-origin and per-browser, so the key does not follow
 * them to their phone - they paste it again there. For a free key that takes a
 * minute to mint, that is the better side of the trade.
 *
 * It is also readable by any script running on this origin, so an XSS bug
 * would expose it. That is true of the session token sitting beside it, and
 * the mitigation is the same: do not have an XSS bug. The blast radius is one
 * revocable, free-tier API key.
 */

/** Deliberately namespaced: this origin also serves the auth cookie and the
 *  Vite dev server, and an unprefixed "apiKey" would be a landmine. */
const STORAGE_KEY = 'resumemaxxer.gemini_api_key'

/**
 * Read the saved key, or null.
 *
 * Wrapped in try/catch because `localStorage` does not merely come back empty
 * in a locked-down context - accessing it *throws*. Safari in private mode and
 * browsers configured to block site data both do this, and an uncaught throw
 * here would take down the whole Tailor page rather than just the key feature.
 */
export function getGeminiKey(): string | null {
  try {
    const stored = window.localStorage.getItem(STORAGE_KEY)
    return stored && stored.trim() ? stored.trim() : null
  } catch {
    return null
  }
}

/** Whether a key is saved. Cheaper to read than to reason about at call sites. */
export function hasGeminiKey(): boolean {
  return getGeminiKey() !== null
}

/**
 * Save a key, or clear it when given something blank.
 *
 * Returns false when storage refused - the caller needs to know, because
 * silently "saving" a key that vanishes on reload is worse than saying no.
 */
export function setGeminiKey(key: string): boolean {
  const trimmed = key.trim()
  try {
    if (!trimmed) {
      window.localStorage.removeItem(STORAGE_KEY)
    } else {
      window.localStorage.setItem(STORAGE_KEY, trimmed)
    }
    return true
  } catch {
    return false
  }
}

/** Forget the saved key. */
export function clearGeminiKey(): void {
  try {
    window.localStorage.removeItem(STORAGE_KEY)
  } catch {
    // Nothing to do: if storage is unavailable there is nothing stored either.
  }
}

/**
 * A masked form for display, e.g. `AIza…9f3K`.
 *
 * The settings UI has to show *something* so the student can tell whether the
 * key they saved is the one they meant, but showing the whole value puts a
 * live credential on screen - in a library, in a screen-share, in a
 * screenshot. First four and last four is enough to recognise a key you chose
 * and useless to anyone who did not.
 */
export function maskGeminiKey(key: string): string {
  const trimmed = key.trim()
  if (trimmed.length <= 12) return '•'.repeat(Math.max(trimmed.length, 4))
  return `${trimmed.slice(0, 4)}…${trimmed.slice(-4)}`
}

/**
 * Catch a mis-paste before spending a request on it.
 *
 * Returns an error message, or null when the key looks plausible. This is not
 * validation - only Google can say whether a key works - it just catches the
 * three mistakes people actually make: pasting half the key, pasting a URL,
 * and pasting with a line break in it.
 *
 * The `AIza` prefix is a WARNING rather than a rejection. Every current AI
 * Studio key starts with it, but hard-rejecting on a prefix would lock out
 * every student the day Google changes the format.
 */
export function describeKeyProblem(key: string): string | null {
  const trimmed = key.trim()

  if (!trimmed) return 'Paste your API key first.'
  if (/\s/.test(trimmed)) {
    return 'That key has a space or line break in it, so it was copied with extra characters. Try again.'
  }
  if (trimmed.startsWith('http')) {
    return 'That looks like a URL, not a key. Copy the key itself from AI Studio.'
  }
  if (trimmed.length < 20) {
    return 'That key looks too short - it may have been cut off. Copy the whole thing.'
  }
  return null
}
